### DISCUSSION TITLE
Events not handled concurrently

### DISCUSSION THREAD

**Discussion by Adam-D-Lewis** [miguelgrinberg/python-socketio#1085] [Q&A]
My issue is very similar to this old issue https://github.com/miguelgrinberg/python-socketio/issues/90, but I'm using uvicorn/FastAPI instead.  I expect the events to be handled concurrently, but they seem to be handled synchronously.

**To Reproduce**
```python
# scratch_server.py
import time
from fastapi import FastAPI
import socketio

app = FastAPI()
app.sio = socketio.AsyncServer(
	async_mode='asgi', 
	cors_allowed_origins='*'
)

app.mount(
	path='/ws', 
	app=socketio.ASGIApp(
		socketio_server=app.sio,
		socketio_path="socket.io",
))

@app.sio.on("ping")
def handle_ping(sid, *args, **kwargs):
    print('got ping')
    time.sleep(1)
    print('returning pong')
    return "pong"
```
```python
# scratch_client.py
import socketio

sio = socketio.Client()
sio.connect(
	"ws://localhost:8000", socketio_path="/ws/socket.io", transports=["websocket"]
)

for i in range(10):
	sio.emit('ping')
```
I launch the server with `uvicorn scratch_server:app --workers 10`.  scratch_client.py takes about 10s to run and the output is:
```
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
got ping
returning pong
```

I expect the events to be handled concurrently and so I expect scratch_client.py to run in about 1s and that I'd see 10 "got ping" logs followed by 10 "returning pong" logs.  Apologies if this is user error, as I am new to this, but any guidance is greatly appreciated :pray: .

**Comment by miguelgrinberg** [comment]
How familiar are you with writing asynchronous code with asyncio? The code that you have written is synchronous, it is invalid for asyncio. First of all, you cannot use `time.sleep` in an asyncio program, you should await the `asyncio.sleep` function instead. Also, your functions should be written as `async def`, not just `def`. I suggest you review the basics of writing asynchronous code and then revisit this.

**Comment by Adam-D-Lewis** [comment]
Thanks for your response, Miguel.  I'm familiar with the basics of asyncio, but there's a lot left I could learn.  FastAPI allows the path operation functions to be sync or async, running the path operation function in a new thread when it's a synchronous function.  I guess I was hoping things would just work, allowing subsequent emitted events from a single client to run in separate threads as well, but that may not be possible with AsyncServer, and I'm assuming socketio.Server isn't compatible with FastAPI since FastAPI implements the ASGI spec.

If anyone else comes across this, I wanted to point out that the following examples do work as I originally intended and might be helpful.

```python
# scratch_server.py
import asyncio
import time

import socketio
from fastapi import FastAPI

app = FastAPI()
app.sio = socketio.AsyncServer(
	async_mode="asgi",
	cors_allowed_origins="*"
)

app.mount(
    path="/ws",
    app=socketio.ASGIApp(
        socketio_server=app.sio,
        socketio_path="socket.io",
    ),
)

@app.sio.on("ping")
async def handle_ping(sid, *args, **kwargs):
    print("got ping")
    # can replace time.sleep with any non-async IO-bound function, and it'll run in a separate thread.  Be careful with race conditions.
    await asyncio.get_event_loop().run_in_executor(None, time.sleep, 1)  
    print("returning pong")
    return "pong"

# also works as originally intended
@app.sio.on("ping2")
async def handle_ping2(sid, *args, **kwargs):
    print("got ping2")
    await asyncio.sleep(1)
    print("returning pong2")
    return "pong2"
```
```python
# scratch_client.py
import socketio

sio = socketio.Client()
sio.connect(
    "ws://localhost:8000", socketio_path="/ws/socket.io", transports=["websocket"]
)

for i in range(10):
    sio.emit("ping")

for i in range(10):
    sio.emit("ping2")
```

**Reply by miguelgrinberg** [reply]
I do not agree with the view that if you have some code that is not asyncio compatible you just send it to an executor and problem solved. Any performance benefit you obtain by using an async application disappears once you introduce executors that create threads and/or processes, so this should be used with caution and avoided if at all possible. But of course, that is my opinion, I'm sure others will disagree. :)

**Reply by synodriver** [reply]
Well, this make me think of [motor](https://github.com/mongodb/motor/blob/6462f1ecbb7e1b39d917f63bf9d6e24fd5cc8735/motor/frameworks/asyncio/__init__.py#L85), the asyncio mongodb driver, which use ```run_in_executor``` and pretend to be async. So it's performance... :)
